### PR TITLE
[subset] update hb_subset api with final changes before going stable.

### PR DESCRIPTION
**Pull request by garretrieger**
Makes some changes to the hb-subset.h API before we stabilize it:

- Remove individual get/set methods for bool properties and instead use an enum of property names.
- hb_subset_input is now const on the hb_subset() call.
- Add hb_subset_or_fail ()
- Adds set/get user data methods that are standard on hb_objects.
- Adds passthrough tables and no subset tables to match fontTools table dropping and retention logic.

I've still got a couple things left to do before this is ready:
- Documentation on the properties enum.
- Documentation for all methods.
- Remove hb_subset() in favor of hb_subset_or_fail()
- Re-order get/set user data methods.

### CODE REVIEW

**Comment by behdad** [comment]
Thanks Garret!

CC @jfkthame @drott @bungeman @mapreri #3012

**Comment by drott** [comment]
Thanks, @garretrieger! API change generally LGTM but I am not too familiar with the API and its usage. `third_party/skia/src/pdf/SkPDFSubsetFont.cpp` owners are better suited to answer.

What's `hb_subset_input_set_user_data` used for? 

IIUC this will need a migration off of `hb_subset_input_set_retain_gids` to `hb_subset_set_flags` in Skia & Chromium. Usually the way that's done is to define a compile flag in Skia, set this compile flag in Chromium's top level `skia/BUILD.gn` `defines` list when rolling HarfBuzz in Chromium. Are you up for handling that migration?

**Comment by behdad** [comment]
> What's `hb_subset_input_set_user_data` used for?

That's generic part of HB object model; so you can attach arbitrary (say, `SkFace`) data to a HB object.

**Comment by behdad** [comment]
> That's generic part of HB object model; so you can attach arbitrary (say, `SkFace`) data to a HB object.

Sometimes used by language bindings.

**Comment by garretrieger** [comment]
> IIUC this will need a migration off of `hb_subset_input_set_retain_gids` to `hb_subset_set_flags` in Skia & Chromium. Usually the way that's done is to define a compile flag in Skia, set this compile flag in Chromium's top level `skia/BUILD.gn` `defines` list when rolling HarfBuzz in Chromium. Are you up for handling that migration?

Yes I can try and do that migration.

File: src/hb-subset-input.cc [hunk 1]
```diff
@@ -237,18 +256,43 @@ hb_subset_input_glyph_set (hb_subset_input_t *subset_input)
   return subset_input->glyphs;
 }
 
+/**
+ * hb_subset_input_nameid_set:
+ * @subset_input: a subset_input.
```
Comment (khaledhosny) [inline src/hb-subset-input.cc:261]: That should be `an #hb_subset_input_t`.

File: src/hb-subset-input.cc [hunk 2]
```diff
@@ -269,98 +313,153 @@ hb_subset_input_get_retain_all_features (hb_subset_input_t *subset_input)
...
+      break;
+  }
 }
 
 /**
- * hb_subset_input_get_retain_gids:
- * Returns: value of retain_gids.
- * Since: 2.4.0
+ * hb_subset_input_set_user_data: (skip)
+ * @face: A subset input object
```
Comment (khaledhosny) [inline src/hb-subset-input.cc:426]: This should be `@input`. There probably should be some consistency in naming this variable, either input or subset_input.

File: src/hb-subset-input.cc [hunk 3]
```diff
@@ -269,98 +313,153 @@ hb_subset_input_get_retain_all_features (hb_subset_input_t *subset_input)
...
 {
-  subset_input->notdef_outline = notdef_outline;
+  return hb_object_set_user_data (input, key, data, destroy, replace);
 }
 
-HB_EXTERN hb_bool_t
-hb_subset_input_get_notdef_outline (hb_subset_input_t *subset_input)
+/**
+ * hb_subset_input_get_user_data: (skip)
+ * @face: A subset input object
```
Comment (khaledhosny) [inline src/hb-subset-input.cc:450]: This should be `@input`.

File: src/hb-subset-plan.cc [hunk 4]
```diff
@@ -389,8 +389,8 @@ _nameid_closure (hb_face_t *face,
  * Since: 1.7.5
  **/
 hb_subset_plan_t *
-hb_subset_plan_create (hb_face_t         *face,
-		       hb_subset_input_t *input)
+hb_subset_plan_create (hb_face_t	 *face,
```
Comment (khaledhosny) [inline src/hb-subset-plan.cc:392]: The function arguments need to be documented.

**Review by khaledhosny — CHANGES_REQUESTED** [review]
I don't think subset documentation is currently generated, appropriate section need to be added to the files unders /docs.

File: src/hb-subset-plan.cc (hunk 4 above)
Comment (garretrieger) [reply src/hb-subset-plan.cc:395]: Done.

File: src/hb-subset-input.cc (hunk 2 above)
Comment (garretrieger) [reply src/hb-subset-input.cc]: Alright, I standardized everything to use "input".

**Comment by behdad** [comment]
Thanks. Garret, can you also add a `SECTION:hb-subset` comment to `hb-subset.cc` to provide some description for this section of the docs? See top of `hb-shape.cc` for example.

**Comment by garretrieger** [comment]
> Thanks. Garret, can you also add a `SECTION:hb-subset` comment to `hb-subset.cc` to provide some description for this section of the docs? See top of `hb-shape.cc` for example.

Done.

**Comment by garretrieger** [comment]
> I don't think subset documentation is currently generated, appropriate section need to be added to the files unders /docs.

Done.

**Review by khaledhosny — APPROVED** [review]

**Pull request merged**